### ISSUE TITLE
[EN] Feedback for: /web/tools/setup/setup-workflow?hl=en

### ISSUE THREAD

**Issue by corysimmons** [google/WebFundamentals#2555]
Why can't Workspaces do very basic stuff like persist classes on DOM elements or persist new properties to preprocessors?

I wrote an [extensive article](https://medium.com/@corysimmons/stop-marketing-your-unfinished-product-2851e15e7427) about this very problem.

**Comment by Garbee** [comment]
It isn't possible to persist edits in the Elements panel back to sources in workspaces. The server sends Chrome the page (HTML.) This is taken, tokenized, then a DOM (Document Object Model) is constructed out of it. The page is then trashed and only the DOM is kept in memory. That is what you are seeing on the Elements panel. That panel is the full DOM as it is, made to look like HTML markup for ease-of-use.

That HTML sent down from the server, can be generated by PHP, Ruby, C, hand-coded, whatever. There is no way for DevTools to know. So, there is no way for us to transfer source edits in the DOM back automatically. CSS and JS both can take advantage of sourcemaps so DevTools know where to send things back. It isn't 100%, since say if you are using a variable you aren't editing the variable back but changing the set value (at least in CSS.) But it is good enough for the majority of uses. With HTML generation by backend languages, these kinds of kinks become even more complex, unmanageable, and will end up providing not the best experience for developers.

Without a "DOM Sourcemap" and "HTML Sourcemap" being done so we can map from the DOM back to HTML, then the HTML back to the generating files, there is no way to handle this. Even then, it becomes an *extremely* complicated problem to solve with the various backend languages and coding styles to adhere to.

The two things we are able to persist, CSS and JavaScript both have a clear mapping system using Sourcemaps. Unfortunately, that is not available nor possible with HTML and the DOM.

**Comment by corysimmons** [comment]
> That HTML sent down from the server, can be generated by PHP, Ruby, C, hand-coded, whatever. There is no way for DevTools to know.

I understand this point. Why not assume it's vanilla HTML and do a 1:1 mapping to a local resource?

> The two things we are able to persist, CSS and JavaScript both have a clear mapping system using Sourcemaps.

Not exactly. I can map a .scss file. It's fine when I edit the pre-existing properties, but if I go to add a new property it adds the new property to the .css rather than the .scss.

**Comment by Garbee** [comment]
> Why not assume it's vanilla HTML and do a 1:1 mapping to a local resource?

Assuming only gets you so far (and people will inevitably be upset.) Also you still have the problem where certain things come from HTML, JS, and more HTML requested by ajax... There is no relationship of *where the DOM comes from* which is the core problem with persisting changes to it.

> I can map a .scss file. It's fine when I edit the pre-existing properties, but if I go to add a new property it adds the new property to the .css rather than the .scss. Not sure if I'm doing something wrong on that one.

If that isn't working I'd file a bug/feature request on [https://crbug.com](crbug). If I recall correctly that *should* work but I don't know with absolute certainty.

**Comment by corysimmons** [comment]
>Assuming only gets you so far (and people will inevitably be upset.)

Assuming with a disclaimer would be better than nothing? It's also interesting that a lot of sites are moving to be completely static:

- https://surge.sh/
- https://www.netlify.com/
- https://www.contentful.com/blog/2015/04/28/webinar-contentful-roots-static-sites/

> If that isn't working I'd file a bug/feature request on crbug If I recall correctly that should work but I don't know with absolute certainty.

I think they're working on it actually.

I guess my point is, what's the point in using Workspaces if it can't do these basic things? I spent a few hours learning all the in's and out's of Workspaces just to discover I couldn't really use it for anything. Why not put a disclaimer on the top of the docs and README that says "This tool is a neat concept but not ready for use" and save people who are looking for a tool time?

**Comment by Garbee** [comment]
As far as technical details on why or why not things should be done, you should file a request on the issue tracker and see what the development team thinks. Nothing more can really be said on those points here since it isn't useful to the documentation efforts.

Instead of asking for us to tell everyone that it is "just a concept" and it is "not ready for use", why not file feature requests on the [issue tracker](https://crbug.com) so that the team that works on the DevTools can see what users needs are. That way they know what more people expect from the tools and they can look into addressing problems developers face in the real world.

The DevTools is heavily influenced by user reports. You can help the team majorly by filing issues that have:

* Clear steps to reproduce the problem.
* Clear action for the team to handle to resolve the issue
* One action per-issue is a major bonus

Issues like that help the team know what users need. If issues get starred by enough users, then that is a strong indicator that people want a certain thing to be carried out. And having one action to an issue to resolve a problem makes it easier to see what is necessary to fix a problem. 5 issues with specific actions even if about a single topic is much better than 1 big issue. That allows for each item to have its own area for technical discussion and starring for letting the team know what is wanted the most.

Help the tooling authors out. Let them know what you expect so they can try to address the need. That way everyone benefits when the tooling is improved to cover more uses.

**Comment by corysimmons** [comment]
Okay I've done so: https://code.google.com/p/chromium/issues/detail?id=586223&thanks=586223&ts=1455217177

You inspired me.
